### ISSUE TITLE
error causes subsequent spurious error about literal cast

### ISSUE THREAD

**Issue by durka** [rust-lang/rust#35772]
See this code:

``` rust
fn main() {
    error;
    1 as f32;
}
```

It has an error on the second line, obviously. But it also prints an error about the third line (illegal cast). But the cast is perfectly legal and removing the first error also "fixes" the second one.

**Comment by durka** [comment]
I suspect this somehow has to do with the constant evaluator. @oli-obk?

**Comment by oli-obk** [comment]
nope, it's not const eval, it's typeck.

beta/nightly even produce "non-scalar cast: `{integer}` as `f32`", so it's not something weird like the `error;` turning the `1` into another type.

**Comment by durka** [comment]
But it _is_ a scalar cast. Maybe the first error leaves some kind of boolean flag around that "there was an error", and the casting code picks up on it?

**Comment by Detegr** [comment]
I've investigated this a bit. Apologies if I've undersood something wrong, I'm still a total compiler noob :)

Looks like the reason this happens is when there's an error before the cast, the type inference of `1` differs from the case when there's just the cast. These types are checked in `src/librustc_typeck/check/cast.rs` in a function called [do_check](https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/check/cast.rs#L312-L335).

In the working version `self.expr_ty.sty` looks like this (from rust-gdb):
```
TyS = {sty = TyInt = {I32}, flags = Cell<rustc::ty::TypeFlags> = {
  value = UnsafeCell<rustc::ty::TypeFlags> = {value = TypeFlags = {
    bits = 0}}}, region_depth = 0}
```
where in the version with the error it looks like this:
```
TyS = {sty = TyInfer = {IntVar = {IntVid = {index = 0}}},
  flags = Cell<rustc::ty::TypeFlags> = {
    value = UnsafeCell<rustc::ty::TypeFlags> = {value = TypeFlags = {
      bits = 3076}}}, region_depth = 0}
```
If `do_check` results in an error (it does if sty is `TyInfer`, see [from_ty](https://github.com/rust-lang/rust/blob/master/src/librustc/ty/cast.rs#L61-L74)) the compiler reports the cast error.

Why the types are resolved differently I have no idea, the code path starting from `self.expr_ty = fcx.structurally_resolved_type(self.span, self.expr_ty);` went way over my understanding.

**Comment by durka** [comment]
Nice detective work @Detegr.

**Comment by Detegr** [comment]
Thanks, however my initial solution to the problem was too shortsighted and caused more harm than good. Anyway this has been a fun dive to the compiler internals and I'll keep trying.

**Comment by estebank** [comment]
[This is fixed in nightly and beta](https://play.rust-lang.org/?gist=6f109acaf64c7018862ba75e77f84766&version=nightly) (probably due to #46732).
